### PR TITLE
refactor: migrate to fastapi

### LINKED ISSUE #3855: [auth] ☂️ migrate to pydantic
- [x] #3863
- [x] #3896
- [x] #3882
- [x] #3900
- [ ] #3903
- [ ] #3904

### PR DESCRIPTION
**Pull request by axiomofjoy**
resolves #3855

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -59,7 +59,8 @@ dependencies = [
   "cachetools",
   "python-multipart",  # see https://www.starlette.io/#dependencies
   "arize-phoenix-evals>=0.13.1",
-  "pyyaml", # for OpenAPI
+  "fastapi",
+  "pydantic>=1.0,<3",
```
Comment (mikeldking) [inline pyproject.toml:63]: I still don't understand why we want this as explicit deps @axiomofjoy - don't we want it as transitive and align with what fastapi deems right?

File: .github/workflows/python-CI.yml [hunk 2]
```diff
@@ -109,6 +109,7 @@ jobs:
             matrix:
                 os: [macos-12, windows-latest, ubuntu-latest]
                 python-version: ["3.8"]
+            fail-fast: false
```
Comment (mikeldking) [inline .github/workflows/python-CI.yml:112]: we sure we want to keep this?

File: pyproject.toml [hunk 3]
```diff
@@ -59,7 +59,8 @@ dependencies = [
   "cachetools",
   "python-multipart",  # see https://www.starlette.io/#dependencies
   "arize-phoenix-evals>=0.13.1",
-  "pyyaml", # for OpenAPI
+  "fastapi",
+  "pydantic>=1.0,!=2.0.*,<3,", # exclude 2.0.* since it does not support the `json_encoders` configuration setting
```
Comment (mikeldking) [inline pyproject.toml:63]: how much does this diverge from fastapi?

**Review by mikeldking — APPROVED** [review]
Approving to unblock. but maybe we let it soak in main for a day or two

File: .github/workflows/python-CI.yml (hunk 2 above)
Comment (axiomofjoy) [reply .github/workflows/python-CI.yml:112]: I personally like this, but will ask the group.

File: pyproject.toml (hunk 3 above)
Comment (axiomofjoy) [reply pyproject.toml:63]: The current version of `fastapi` officially has a lower bound of `1.7.4`, but it does not actually appear to be compatible with that version. See comment [here](https://github.com/Arize-ai/phoenix/issues/3855#issuecomment-2244036126).

**Pull request merged**